packages/amplify_api/lib/src/graphql/graphql_request_factory.dart

```diff
@@ -0,0 +1,135 @@
+
+class GraphQLRequestFactory {
+  String _getModelType(ModelFieldTypeEnum? val) {
+    switch (val) {
+      case ModelFieldTypeEnum.string:
+        return "String";
+      case ModelFieldTypeEnum.int:
+        return "int";
+      case ModelFieldTypeEnum.model:
+        return "ID";
+      default:
+        return "Error: UnknownType!";
+    }
+  }
+
+  String _getFieldsFromModelType(ModelSchema schema) {
+    Map<String, ModelField> fieldsMap = schema.fields!;
+    return fieldsMap.entries
+        .map((entry) => entry.value.association == null ? entry.key : '')
+        .toList()
+        .join(' ');
+  }
+
+  ModelSchema _getAndValidateSchema(
+      ModelType modelType, GraphQLRequestOperation operation) {
+    ModelProviderInterface? provider = AmplifyAPI.instance.getModelProvider();
+
+    if (provider == null) {
+      throw ApiException('No modelProvider found',
+          recoverySuggestion:
+              'Pass in a modelProvider instance while instantiating APIPlugin');
+    }
+
+    ModelSchema schema = provider.modelSchemas.firstWhere(
+        (elem) => provider.getModelTypeByModelName(elem.name) == modelType,
+        orElse: () => throw ApiException(
+            'No schema found for the ModelType provided',
+            recoverySuggestion:
+                'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType'));
+
+    if (schema.fields == null) {
+      throw ApiException('Schema found does not have a fields property',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin');
+    }
+
+    if (operation == GraphQLRequestOperation.list &&
+        schema.pluralName == null) {
+      throw ApiException('No schema name found',
+          recoverySuggestion:
+              'Pass in a valid modelProvider instance while instantiating APIPlugin or provide a valid ModelType');
+    }
+
+    return schema;
+  }
+
+  /*
+   *  buildQuery()
+   *
+   *  
+   *  Example: 
+   *    query getBlog($id: ID!) { getBlog(id: $id) { id name createdAt } }
+  */
+  GraphQLRequest<T> buildQuery<T extends Model>(
+      {required ModelType modelType,
+      required Map<String, String>? variableInput,
+      required String? id,
+      required GraphQLRequestType requestType,
+      required GraphQLRequestOperation requestOperation}) {
+    ModelSchema schema = _getAndValidateSchema(modelType, requestOperation);
+
+    String name = schema.name;
+
+    // fields to retrieve, e.g. "id name createdAt"
+    String fields = _getFieldsFromModelType(schema);
+
+    // e.g. "query"
+    String requestTypeVal = describeEnum(requestType.toString());
+    // e.g. "get"
+    String requestOperationVal = describeEnum(requestOperation.toString());
+
+    // Upper document input: ($id: ID!)
+    var varInputUpperStr = '';
+    // Lower document input: (id: $id)
+    var varInputLowerStr = '';
+
+    if (variableInput != null && variableInput.isNotEmpty) {
```

Comment:
The input should not be used for building the query. Only the schema and the `_getModelType` function are required. For one, this variable input is not being validated against the schema and two, the values of the map cannot serve both as the type parameter and the value of the argument. For example, for a query op, `variableInput` should be `{'id': model.id}`, and the field name and type should be retrieved from the model schema.

Reply:
You can use `ModelFieldDefinition.idKey` for retrieving the primary key field, which may or may not be of type `ID!`. See [here](https://docs.amplify.aws/cli/graphql-transformer/key#example-get-customers-by-email) where `email: String!` is the primary key.

Reply:
I've refactored this section. It should now read clearer and be more generic so we can reuse down the road. Thanks for the insight 💯